### PR TITLE
Separate idnr input

### PR DESCRIPTION
**Pull request by frederike-ramin**
# Short Description
- This splits the input field of the idnr into four parts

# Changes
- Make the IdNrField
- Remove details/help where not necessary (person a and person b)
- Fix the appearance of form spacing. (previously, I matched too many elements, including all headings etc when I only wanted form-rows to be matched.)

# Feedback
- I am still talking to Nadine whether or not the implementation of the idnr input is okay (you input one digit into every field, click submit -> they appear in pairs of two in the first two fields. Do you have another intuition on how we could possibly avoid this? (I thought about adding separating characters as with the entries field, it just fields awkward to me to store the idnr as a list and I guess this would also change a lot in the code and we'd have to be very careful. Do you have an opinion/idea?

### CODE REVIEW

**Comment by frederike-ramin** [comment]
I talked to Nadine and I'm redoing the usability-part.

**Comment by frederike-ramin** [comment]
Alright, I think I really like the solution, but let me know what you think. The solution relies on the fact, that we only have validated data in our stored data after we once validated the form. The usability problem described above afterall only occurred in the case your input for the idnr was wrong. - I think. Maybe you can reevaluate that. I tried my best to bust the solution and couldn't do it^^

File: acceptance_tests/cypress/integration/lotse_flow.spec.js [hunk 1]
```diff
@@ -347,8 +368,8 @@ context('Acceptance tests', () => {
                 cy.visit('/lotse/step/summary')
                 cy.contains('Sie müssen eingeloggt sein')
                 cy.visit('/unlock_code_activation/step/data_input')
-                cy.get('#idnr').should('have.value', '')
-                cy.get('#unlock_code').should('have.value', '')
+                cy.get('#idnr_1').should('have.value', '')
```
Comment (JulianRoesner) [inline acceptance_tests/cypress/integration/lotse_flow.spec.js:371]: Shouldn't all four field have value ''?

File: webapp/app/forms/fields.py [hunk 2]
```diff
@@ -63,6 +63,46 @@ def _value(self):
...
+
+        # Once the validation has gone through, post_validate() stores the data as string.
+        # As we know that it is correct, we can just separate it in chunks here.
+        split_data = []
+        chunk_sizes = self.widget.input_field_lengths
+        current_idx = 0
+        for chunk_size in chunk_sizes:
+            if self.data:
+                split_data.append(self.data[current_idx: current_idx + chunk_size])
+            current_idx += chunk_size
```
Comment (JulianRoesner) [inline webapp/app/forms/fields.py:96]:
```suggestion
        start_idx = 0
        for chunk_size in chunk_sizes:
            end_index = start_idx + chunk_size
            if self.data:
                split_data.append(self.data[start_idx: end_index])
            start_idx = end_index
```

I was just thinking how to make it more readable. However, it is not that important

File: webapp/app/forms/fields.py [hunk 3]
```diff
@@ -63,6 +63,46 @@ def _value(self):
...
+            self.data = valuelist
+        elif self.data is None:
+            self.data = []
+
+    def _value(self):
+        # In case the validation was not successful, we already have the data as a list of values.
+        if isinstance(self.data, list):
+            return self.data
+
+        # Once the validation has gone through, post_validate() stores the data as string.
```
Comment (JulianRoesner) [inline webapp/app/forms/fields.py:88]: So _value is only called after post_validate has been called?

File: webapp/app/forms/fields.py [hunk 4]
```diff
@@ -63,6 +63,46 @@ def _value(self):
...
+    def process_formdata(self, valuelist):
+        if valuelist:
+            self.data = valuelist
+        elif self.data is None:
+            self.data = []
+
+    def _value(self):
+        # In case the validation was not successful, we already have the data as a list of values.
+        if isinstance(self.data, list):
+            return self.data
```
Comment (JulianRoesner) [inline webapp/app/forms/fields.py:86]: I do not really get how that is only the case for unsuccessful validations?

**Review by JulianRoesner — COMMENTED** [review]
I think I'll need a little more explanation to some of the parts. Nevertheless, it looks very good to me.

File: acceptance_tests/cypress/integration/lotse_flow.spec.js (hunk 1 above)
Comment (frederike-ramin) [reply acceptance_tests/cypress/integration/lotse_flow.spec.js:371]: Yeees..I honestly just was too lazy. 🙈

File: webapp/app/forms/fields.py (hunk 3 above)
Comment (frederike-ramin) [reply webapp/app/forms/fields.py:88]: Mh..no..the comments aren't clear enough, I see^^

Comment (frederike-ramin) [reply webapp/app/forms/fields.py:102]: How about now?

File: webapp/app/forms/fields.py [hunk 5]
```diff
@@ -70,18 +70,32 @@ class IdNrWidget(MultipleInputFieldWidget):
...
     def process_formdata(self, valuelist):
+        # The formdata (from the request) is written to self.data as is (as a list of inputted strings)
         if valuelist:
             self.data = valuelist
         elif self.data is None:
             self.data = []
 
     def _value(self):
-        # In case the validation was not successful, we already have the data as a list of values.
+        """ Returns the representation of data as needed by the widget. In this case: a list of strings. """
```
Comment (JulianRoesner) [inline webapp/app/forms/fields.py:96]: This is always only called after a successful validation, correct?

**Review by JulianRoesner — APPROVED** [review]
That is very clear now, thanks

File: webapp/app/forms/fields.py (hunk 5 above)
Comment (frederike-ramin) [reply webapp/app/forms/fields.py:96]: No, it is not. This is the method that is called whenever the form wants to know what to display in the widget. So whenever the field is rendered, this is called.

Comment (JulianRoesner) [reply webapp/app/forms/fields.py:96]: Okidoki

**Pull request merged**